### ISSUE TITLE
space-before-function-paren reports spurious error with async

### ISSUE THREAD

**Issue by nirvdrum** [eslint/eslint#13211]
**Tell us about your environment**

Environment Info:

Node version: v12.16.1
npm version: v6.14.3
Local ESLint version: v6.8.0 (Currently used)
Global ESLint version: Not found

**What parser (default, Babel-ESLint, etc.) are you using?**

`@typescript-eslint/parser`

**Please show your full configuration:**

<details>
<summary>Configuration</summary>

```js
module.exports = {
  parser: '@typescript-eslint/parser',
  extends: [
    'eslint:recommended',
    'plugin:react/recommended',
    'plugin:react-hooks/recommended',
    'plugin:relay/recommended',
    'plugin:@typescript-eslint/eslint-recommended',
    'plugin:@typescript-eslint/recommended',
    'prettier',
    'prettier/@typescript-eslint',
    'prettier/react',
  ],
  plugins: ['relay'],
  parserOptions: {
    ecmaFeatures: {
      jsx: true, // Allows for the parsing of JSX
    },
  },
  rules: {
    '@typescript-eslint/ban-ts-ignore': 'off',
    '@typescript-eslint/camelcase': 'off',
    '@typescript-eslint/explicit-function-return-type': 'off',
    '@typescript-eslint/explicit-member-accessibility': [
      'error',
      {
        accessibility: 'no-public',
      },
    ],
    '@typescript-eslint/no-explicit-any': 'off',
    '@typescript-eslint/no-inferrable-types': 'off',
    '@typescript-eslint/no-non-null-assertion': 'off',
    '@typescript-eslint/no-unused-vars': 'off',
    '@typescript-eslint/no-use-before-define': 'off',
    'react/display-name': 'off',
    'react/no-unescaped-entities': 'off',
    'react/prop-types': 'off',
    'relay/generated-flow-types': 'off',
    curly: 'error',
    'space-before-function-paren': ['error', 'never'],
  },
  settings: {
    react: {
      version: 'detect',
    },
  },
}

```

</details>

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

```js
    const asyncLogin = async () => {
        await context.loginWithRedirect({ appState: { targetUrl: window.location.pathname } })
    }
```

```bash
yarn eslint --ignore-path .gitignore './**/*.{js,ts,tsx}'
```

**What did you expect to happen?**

I expected the _space-before-function-paren_ to detect call sites where a function name is separate from its argument list by whitespace.

**What actually happened? Please include the actual, raw output from ESLint.**

The rule appears to think `async` is a function name and reports the sample call I made as an error:

```
20:29  error  Unexpected space before function parentheses  space-before-function-paren
```

Instead, the rule should realize that `async` is a keyword used to modify the anonymous function declaration that follows it.

**Are you willing to submit a pull request to fix this bug?**

If there's agreement that the case I've indicated is in error, I can take a pass at fixing it. I've never worked with the ESLint codebase before, however, and would need to familiarize myself with the AST in order to adequately fix this.

**Comment by anikethsaha** [comment]
I think its the expected behavior. cause this rule focus on `parens` for the function arguments. it expects space before it whether its function name or async or anything else, it needs space before `(`.

you could try with  `{"asyncArrow": "always"}`.

[demo](https://eslint.org/demo#eyJ0ZXh0IjoiLyplc2xpbnQgc3BhY2UtYmVmb3JlLWZ1bmN0aW9uLXBhcmVuOiBbXCJlcnJvclwiLCB7XCJhbm9ueW1vdXNcIjogXCJuZXZlclwiLCBcIm5hbWVkXCI6IFwibmV2ZXJcIiwgXCJhc3luY0Fycm93XCI6IFwiYWx3YXlzXCJ9XSovXG5jb25zdCBhc3luY0xvZ2luID0gYXN5bmMgKCkgPT4ge1xuICAgICAgICBhd2FpdCBjb250ZXh0LmxvZ2luV2l0aFJlZGlyZWN0KHsgYXBwU3RhdGU6IHsgdGFyZ2V0VXJsOiB3aW5kb3cubG9jYXRpb24ucGF0aG5hbWUgfSB9KVxuICAgIH0iLCJvcHRpb25zIjp7InBhcnNlck9wdGlvbnMiOnsiZWNtYVZlcnNpb24iOjExLCJzb3VyY2VUeXBlIjoic2NyaXB0IiwiZWNtYUZlYXR1cmVzIjp7fX0sInJ1bGVzIjp7InNwYWNlLWJlZm9yZS1mdW5jdGlvbi1wYXJlbiI6Mn0sImVudiI6e319fQ==)

let me know if I am missing something.

**Comment by nirvdrum** [comment]
Apologies if I wasn't clear. In my case, I don't want there ever to be spaces between the function name and the argument list. My config marks it as `"never"`. The issue is the rule is treating `async` as a function name, when it isn't. In the example I gave (reproduced here):

```
const asyncLogin = async () => {
    await context.loginWithRedirect({ appState: { targetUrl: window.location.pathname } })
}
```

The rule is not satisfied until I change the code to:

```
const asyncLogin = async() => {
    await context.loginWithRedirect({ appState: { targetUrl: window.location.pathname } })
}
```

There is no function name here, as it's an anonymous function, so the rule shouldn't be activating. Moreover, `async` is a keyword, not a function name. I don't think forcing people to write `async()` (without a space) is what anyone would expect when activating this rule.

**Comment by anikethsaha** [comment]
sorry, I missed `no space` in my comment , I changed it.

```js
const asyncLogin = async () => {
    await context.loginWithRedirect({ appState: { targetUrl: window.location.pathname } })
}
```
with option `never` is error according to the rule. it checks for space before `(`.

so 
- never
```js
const asyncLogin = async () => { //  space before `(` so error 
    await context.loginWithRedirect({ appState: { targetUrl: window.location.pathname } })
}
```

- always
```js
const asyncLogin = async () => { //  lint free as space before `(`
    await context.loginWithRedirect({ appState: { targetUrl: window.location.pathname } })
}
```

I will check the source once again 👍 , may be I am wrong

**Comment by nirvdrum** [comment]
Okay. Just to reiterate... `async` isn't the function name. It shouldn't be pertinent to either the function definition (other than as a modifier) or invocation.

**Comment by anikethsaha** [comment]
this is intentional. it checks for any token irrespective of type (except for whitespace) before the opening paren. 

mainly if there is a space between first token and opening paren for `async` function, it will report if the option  is `never` and if option if `always` then it wont .

**Comment by nirvdrum** [comment]
I've never seen anyone write an anonymous async function as `async() => {}`. That makes it look like you're calling a function named `async` rather than modifying an anonymous function with the `async` keyword. I appreciate that's what the rule does today, but I think that defies most user expectations and is thus a bug.

I, as a user, expected this rule to apply in two contexts:

* Following a named function definition (e.g., `function myFunction(x, y)`)
* Following a function invocation, either from a named function or a captured reference to a function

I say this as someone that never inserts spaces between the function name and the parameter or argument list when writing JavaScript (i.e., I'd never call `myFunction (x, y)`. I'd like to activate this rule, but being forced to have to write `async() => {}` is so unnatural that I'd rather just rely on my team's ability to not violate the rule without an explicit check.

What I'm proposing is a more nuanced application of this rule that looks at the sequence of AST nodes and doesn't activate for `async`.

**Comment by anikethsaha** [comment]
ohh, so you are proposing a rule change request right? I thought you are reporting bug as there is `bug` labeled (from the template).

**Comment by mdjermanovic** [comment]
@nirvdrum thanks for the issue!

`space-before-function-paren` requires or disallows space before the opening paren of parameters, and can be configured separately for named functions (space between the function's name and the paren), anonymous functions (space between `function` and the paren) and async arrow functions (space between `async` and the paren). 

Please see the [Options](https://eslint.org/docs/rules/space-before-function-paren#options) section in the docs.

You would probably want a configuration like this:

```js
/*eslint space-before-function-paren: ["error", {
    "named": "never",
    "anonymous": "always",    
    "asyncArrow": "always"
}]*/

function foo() {}; // ok

const bar = async () => {}; // ok

const baz = function () {}; // ok
```

[Online Demo](https://eslint.org/demo#eyJ0ZXh0IjoiLyplc2xpbnQgc3BhY2UtYmVmb3JlLWZ1bmN0aW9uLXBhcmVuOiBbXCJlcnJvclwiLCB7XG4gICAgXCJuYW1lZFwiOiBcIm5ldmVyXCIsXG4gICAgXCJhbm9ueW1vdXNcIjogXCJhbHdheXNcIiwgICAgXG4gICAgXCJhc3luY0Fycm93XCI6IFwiYWx3YXlzXCJcbn1dKi9cblxuZnVuY3Rpb24gZm9vKCkge307IC8vIG9rXG5cbmNvbnN0IGJhciA9IGFzeW5jICgpID0+IHt9OyAvLyBva1xuXG5jb25zdCBiYXogPSBmdW5jdGlvbiAoKSB7fTsgLy8gb2siLCJvcHRpb25zIjp7InBhcnNlck9wdGlvbnMiOnsiZWNtYVZlcnNpb24iOjExLCJzb3VyY2VUeXBlIjoic2NyaXB0IiwiZWNtYUZlYXR1cmVzIjp7fX0sInJ1bGVzIjp7fSwiZW52Ijp7fX19)

For function invocations, there's another rule: [func-call-spacing](https://eslint.org/docs/rules/func-call-spacing)

**Comment by nirvdrum** [comment]
@mdjermanovic Thanks for clarifying. @anikethsaha had previously mentioned "asyncArrow", but I understood that to be its own rule and it naturally didn't work. I also apologize if I picked the wrong issue template. I took this to be a usability issue and couldn't believe anyone would expect `async()` to be enforced, so I incorrectly presumed this to be a bug.

It looks like this is my fault for not paying close enough to the documentation, so I'm sorry for any confusion. I'm still pretty surprised that there's a separate config item for `async()` vs `async ()`. Semantically, it looks entirely different from the other two cases. Setting aside that you can't have anonymous functions in these cases, I think `async` falls into the same class of functino modifiers as `static`, `set`, `get`, and visibility modifiers, and not in the same class as `function`, which is used to declare a function.

I still strongly suspect that anyone using `space-before-function-paren: ["error", "never"]` is not going to expect the async arrow case to come into play, but I can appreciate why you wouldn't want the short-hand rule to apply different to each sub-case. I guess I'd advocate for "asyncArrow" to be removed, but presumably someone finds value in having it, so I won't file a rule change for it. I'll adjust my ESLint config accordingly. Thanks!
